### PR TITLE
feat: support externally signed credentials (just-in-time data)

### PR DESCRIPTION
**Pull request by nanderstabel**
# Description of change
Adds support for just-in-time external signed credentials. 

With this change, UniCore can be configured such that (signed) credentials can be supplied from an external source just-in-time (during a OpenID4VCI flow). 

Instead of first posting a credential to the `/v1/credentials` endpoint, and then `/v1/offers`, the Http Event Publisher can be configured to listen to `CredentialRequestVerified` events. The event (which includes the `offer_id` and `subject_id`) will be dispatched to the configurated external server endpoint where the `subject_id` can be included in a digital credential which can then signed and send back to the `/v1/credentials` endpoint.

For more information see the update in `agent_application/docker/README.md`

Other notable changes:
- In order for Event Publishers to be connected to events in `agent_issance`, The `fn issuance_state` functions in  `in_memory.rs` and `postgres.rs` had to be updated in a similar fashion as their `fn verification_state` counterparts (`agent_verification` events could already be dispatched by the Event Publisher)
- The `subjectId` is renamed to `offerId` 
- Signing of credentials was done in the `Offer` aggregate. this functionality is moved to the `Credential` aggregate instead.

**Blocked by**
- [x] #44 

## Links to any relevant issues
- #23 

## How the change has been tested
Tested by updating the `test_credential_endpoint` test with three test cases:
1. without an external server (this is the flow as it already existed
2. with an external server (credential is not externally signed and thus will be signed in UniCore)
3. with an external server and self signed (externally signed) credential (UniCore will not do the signing)

## Definition of Done checklist
Add an `x` to the boxes that are relevant to your changes.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: agent_api_rest/src/issuance/credential_issuer/credential.rs [hunk 1]
```diff
@@ -60,13 +43,58 @@ pub(crate) async fn credential(
...
         credential_request,
-        credentials,
     };
 
+    // Use the `offer_id` to verify the `proof` inside the `CredentialRequest`.
+    if command_handler(&offer_id, &state.command.offer, command).await.is_err() {
+        StatusCode::INTERNAL_SERVER_ERROR.into_response();
+    };
+
+    std::thread::sleep(std::time::Duration::from_millis(100));
```
Comment (daniel-mader) [inline agent_api_rest/src/issuance/credential_issuer/credential.rs]: We should extract this magic number as a `const`

File: agent_application/docker/README.md [hunk 2]
```diff
@@ -45,3 +45,31 @@ docker compose up
 
 > [!NOTE]
 > If you don't have rewrite rules enabled on your reverse proxy, you can set the `AGENT_CONFIG_BASE_PATH` to a value such as `ssi-agent`.
+
+## Leveraging Just-in-Time Data Request Events
+
+UniCore empowers dynamic integration with external systems through just-in-time data request events. By configuring an HTTP Event Publisher, events can seamlessly dispatch to external systems, enabling systematic integration with existing infrastructures. This feature proves invaluable in scenarios requiring real-time data retrieval from external sources or on-demand data generation.
+
+By leveraging just-in-time data request events, you enhance the flexibility and efficiency of data management within your SSI ecosystem. UniCore seamlessly integrates with external systems, facilitating on-demand data access and enriching the versatility of your infrastructure.
```
Comment (daniel-mader) [inline agent_application/docker/README.md]: `[self-resolvable]`
I feel like this is a little repetitive of the previous paragraph. Maybe the two paragraphs can be merged into one.

File: agent_application/docker/README.md [hunk 3]
```diff
@@ -45,3 +45,31 @@ docker compose up
...
+
+### Practical Applications
```
Comment (daniel-mader) [inline agent_application/docker/README.md]: `[self-resolvable]`
Rename to `Example scenarios`?

**Review by daniel-mader — DISMISSED** [review]

File: agent_api_rest/src/issuance/credential_issuer/credential.rs (hunk 1 above)
Comment (nanderstabel) [reply agent_api_rest/src/issuance/credential_issuer/credential.rs]: Thanks for reminding me :ok_hand:

**Review by daniel-mader — APPROVED** [review]

**Pull request merged**